src/lib/palette.tsx

```diff
@@ -0,0 +1,42 @@
+import { css } from 'styled-components';
+
+/**
+ *
+ * @param color : hsl => "255, 85%, 58%"";
+ */
+
+export const getHue = (color: string) => {
+  const colors = color.replace(' ', '').split(',');
+  const hue = Number.parseInt(colors[0], 10);
+  const saturation = Number.parseFloat(colors[1]);
+  const lightness = Number.parseFloat(colors[2]);
+  return { hue, saturation, lightness };
+};
+
+const lightnessFn = (number: number, saturation: number) => saturation + 50 - 0.1 * number;
+
+const saturationFn = (number: number, saturation: number) => {
+  /** for black and white palette => saturation doesn't change */
+  if (saturation < 30 || number === 500) return saturation;
+  if (number > 500) return saturation - 30;
+  return saturation - 10;
+};
+
+const COLOR_NUMBERS = [100, 200, 300, 400, 500, 600, 700, 800];
+
+export const createPalette = (color: string, name: string) => {
+  const { hue, saturation, lightness } = getHue(color);
+
+  const palette = COLOR_NUMBERS.map((number) => ({
+    name: `${name}-${number}`,
+    hue,
+    saturation: saturationFn(number, saturation),
+    lightness: lightnessFn(number, lightness),
+  }));
+
+  return css`
+    ${palette.map(
+      ({ name, hue, saturation, lightness }) => `--${name}: ${hue}, ${saturation}%, ${lightness}%;`,
+    )}
+  `;
+};
```

Comment:
Возможно, это можно и в одной итерации сделать. Пример кода сделан с учетом переименования `getHue` на `getHSL`. 
Наверное, от этого постарадает читаемость (??)
```suggestion
export const createPalette = (color: string, name: string) => {
  const hsl = getHSL(color);

  const palette = COLOR_NUMBERS.map((number) => {
    const colorName = `${name}-${number}`;
    const saturation = saturationFn(number, hsl.saturation);
    const lightness = lightnessFn(number, hsl.lightness);

    return `--${colorName}: ${hsl.hue}, ${saturation}%, ${lightness}%;`;
  });

  return css`
    ${palette}
  `;
};
```